homeassistant/components/honeywell/__init__.py

```diff
@@ -107,17 +107,24 @@ async def _retry(self) -> bool:
         if self._client is None:
             return False
 
-        devices = [
+        refreshed_devices = [
             device
             for location in self._client.locations_by_id.values()
             for device in location.devices_by_id.values()
         ]
 
-        if len(devices) == 0:
-            _LOGGER.error("Failed to find any devices")
+        if len(refreshed_devices) == 0:
+            _LOGGER.error("Failed to find any devices after retry")
             return False
 
-        self.devices = devices
+        for configured_device in self.devices:
+            for refreshed_device in refreshed_devices:
+                if (
+                    configured_device.deviceid == refreshed_device.deviceid
+                    and configured_device.name == refreshed_device.name
+                ):
+                    configured_device = refreshed_device
```

Comment:
I'm 95% certain there is a better way to do this. The only thing in the way of making this more beautiful is my very limited knowledge of python. 

Would love to hear of all the better ways of doing this out there. 👍

Reply:
Is this trying to update self.devices by iterating over the list and updating the element when it finds a match? I don't think this works, if I understand:
```
>>> l = ['a', 'b', 'c']
>>> for x in l:
...   if x == 'b':
...     x = 'g'
>>> l
['a', 'b', 'c']
```

Where exactly are the devices getting added with the same id? My initial thought is to try to keep a dictionary based on the device id, but I'm having trouble following the logic of this file.

Reply:
You're correct. I have an array of `device` objects that I want to iterate through and change to the new updated `device` objects from a different array (line 110). Now that you mention it, yes I think a dictionary makes sense, but it may require some sort of migration for users who already have the honeywell integration.

Reply:
Why is that? This looks like data only handled at runtime, and isn't persisted anywhere.

Reply:
Updated the integration to store devices in a dictionary.